File: src/algorithms/01_sorting/04_quicksort/implementations/ts/quick_sort.ts
export function quick_sort(unsortedArray: Array<number>): Array<number> {
  if (unsortedArray.length <= 1) {
    return unsortedArray;
  }

  const { pivotIndex, partitionedArray } = partition(unsortedArray);
  const left = partitionedArray.slice(0, pivotIndex);
  const right = partitionedArray.slice(pivotIndex);

  const leftSorted = quick_sort(left);
  const rightSorted = quick_sort(right);

  return leftSorted.concat(rightSorted);
}

export function partition(unsortedArray: Array<number>): {
  pivotIndex: number;
  partitionedArray: Array<number>;
} {
  const array = [...unsortedArray];
  let i = 0; // start pointer on first item
  let right = array.length - 1; // last item

  // from pointer to item before last
  for (let j = 0; j < right; j++) {
    // if current item is smaller than pivot, swap it with item under pointer and increase pointer
    if (array[j] <= array[right]) {
      [array[i], array[j]] = [array[j], array[i]];
      i++;
    }
  }

  // Center the pivot between lesser and greater items by swapping it with first greater item
  [array[i], array[right]] = [array[right], array[i]];

  return {
    pivotIndex: i,
    partitionedArray: array,
  };
}

File: src/algorithms/01_sorting/04_quicksort/implementations/rs/quick_sort.rs
pub fn quick_sort(unsorted_array: Vec<i32>) -> Vec<i32> {
  let cloned_array = unsorted_array.clone();

  if unsorted_array.len() <= 1 {
    return unsorted_array;
  }

  let (pivot_index, partitioned_array) = partition(cloned_array);
  let left = partitioned_array[..pivot_index].to_owned();
  let right = partitioned_array[pivot_index..].to_owned();

  let left_sorted = quick_sort(left);
  let right_sorted = quick_sort(right);

  [left_sorted, right_sorted].concat()
}

pub fn partition(unpartitioned_array: Vec<i32>) -> (usize, Vec<i32>) {
  let mut cloned_array = unpartitioned_array.clone();
  let mut i = 0;

  let right = cloned_array.len() - 1;

  for j in 0..right {
    if cloned_array[j] <= cloned_array[right] {
      cloned_array.swap(j, i);
      i += 1;
    }
  }

  cloned_array.swap(i, right);

  (i, cloned_array)
}
